[ADD] estate: init commit for estate module

ybenbrai opened the pull request.

ybenbrai committed:
[ADD] estate: init commit for estate module

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1027.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1027)

ybenbrai committed:
[IMP] estate: database field creation

added more fields to the database table with needed requirements

ybenbrai committed:
[IMP] estate: access rights

added access rights

ybenbrai committed:
[IMP] estate: fixing linting issues

	fixed the issues on linting

ybenbrai committed:
[IMP] estate: Creating ui action

actions menu view

clbr-odoo commented on estate/security/ir.model.access.csv:
This should be in a folder named `security` by convention, not `data`.

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,35 @@
+from odoo import fields, models
+
+
+class Property(models.Model):
```
```suggestion
class EstateProperty(models.Model):
```
There is no big incidence here but we generally put the name of the file in CamelCase here.
It's usually good to be specific. `Property` is super generic and could be existing in other modules.

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,35 @@
+
+class Property(models.Model):
+    _name = "estate.property"
+    _description = "This is a estate property"
```
```suggestion
    _description = "Estate property"
```

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,35 @@
+class Property(models.Model):
+    _name = "estate.property"
+    _description = "This is a estate property"
+    _order = "sequence"
```
That `_order` will define how the `search` will return you the records.
You don't have a `sequence` field on your `estate.property` model, leave it empty to keep the default (that order by ID) for now.

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,35 @@
+    _description = "This is a estate property"
+    _order = "sequence"
+
+    name = fields.Char('Name', required=True)
```
```suggestion
    name = fields.Char(string='Name', required=True)
```
Note that what you did is totally equivalent to this ☝️ . Some reviewers prefer that we name all parameters explicitly.
Also note that by default, Odoo will use the variable name to create a label if not provided.
In this case it will create "Name", so you don't really need to specify it.

ybenbrai replied:
> Note that what you did is totally equivalent to this ☝️ . Some reviewers prefer that we name all parameters explicitly. Also note that by default, Odoo will use the variable name to create a label if not provided. In this case it will create "Name", so you don't really need to specify it.

okay noted !, i changed it to `Title` in local now to match the screenshots

clbr-odoo commented on estate/view/__init__.py:
1. You don't need to have an `__init__` file in your `view` folder as it does not contain Python files.
2. By convention, that folder should not be `view` but `views`. Please double-check the [coding guidelines](https://www.odoo.com/documentation/19.0/contributing/development/coding_guidelines.html#directories) for reference. Also don't hesitate to start to explore Odoo codebase and copy how it's done 🙂

clbr-odoo commented on estate/view/estate_property_views.xml:
```diff
@@ -0,0 +1,18 @@
+              action="estate_property_action" 
+              parent="estate_menu_advertisements"/>
+
+</odoo>
```
Missing EOL at the end of the file.

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,17 @@
+    ],
+     'data': [
+         'data/ir.model.access.csv',
+         'view/estate_property_views.xml',
```
```suggestion
         'security/ir.model.access.csv',
         'views/estate_property_views.xml',
```

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,17 @@
+    'summary': 'Track Real Estate',
+    'website': 'https://www.odoo.com/app/realestate',
+    'author': 'Odoo S.A.',
+    'license': 'LGPL-3',
```
```suggestion
    'application': True,
    'license': 'LGPL-3',
```
You probably want to have application=True for this module. See the [manifest reference](https://www.odoo.com/documentation/19.0/developer/reference/backend/module.html).

clbr-odoo commented on estate/view/estate_property_views.xml:
```diff
@@ -0,0 +1,18 @@
+        <field name="view_mode">tree,form</field>
+    </record>
+
+    <menuitem id="estate_menu_root" name="Real Estate"/>
```
Note that you can also nest the module to avoid defining the `parent`, like this:
```xml
<menuitem ...>
    <menuitem .../>
</menuitem/>
```

clbr-odoo reviewed (commented):
Good job so far, just a few comments ! 

Your commit title are not fully correct: usually we put the `[ADD]` tag on the module init (where the manifest is). Following commits will be `[IMP](provements)` to the now existing module.
See [git guidelines](https://www.odoo.com/documentation/19.0/developer/reference/backend/module.html) for more details.

ybenbrai committed:
[IMP] estate: fix comments and apply suggestions

ybenbrai committed:
[IMP] estate: improvement to the state form

added improvement to the properties

ybenbrai committed:
[IMP] estate: fixing styling issues

ybenbrai committed:
[IMP] estate: adding features to the app

added search / filter / form / header

ybenbrai committed:
[FIX] estate: fix for ci checks

	fixed the issue for the styles and the tutorial syntax

antonrom1 commented on estate/views/estate_tree.xml:
```diff
@@ -0,0 +1,17 @@
+            </list>
+        </field>
+    </record>
+</odoo>
```
There is a missing end of line here

ybenbrai committed:
[IMP] estate: Adding features to the estate app

	added tags types.. and accesses and infos in the form

ybenbrai committed:
[IMP] estate: added feature to compute fields

	Chapter 8: Computed Fields And Onchanges
	count the oriontation depend of criteria using action

ybenbrai committed:
[FIX] estate: fixed style issues

ybenbrai committed:
[IMP] estate: added accept/refuse offer

	added requirement of chapter 9

ybenbrai committed:
[FIX] estate: Fixed ci style & tutorial issues

clbr-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,5 @@
+access_estate_property,access.estate.property,model_estate_property,base.group_user,1,1,1,1
+access_estate_property_type,access.estate.property.type,model_estate_property_type,base.group_user,1,1,1,1
+access_estate_property_tag,access.estate.property.tag,model_estate_property_tag,base.group_user,1,1,1,1
+access_estate_property_offer,access.estate.property.offer,model_estate_property_offer,base.group_user,1,1,1,1
```
EOL missing at the end of the file.

clbr-odoo commented on estate/views/estate_list.xml:
```diff
@@ -0,0 +1,18 @@
+            </list>
+        </field>
+    </record>
+</odoo>
```
EOL

clbr-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,11 @@
+            <menuitem id="estate_property_tag_menu_action" name="Property Tags" action="estate_property_tag_action" />
+        </menuitem>
+    </menuitem>
+</odoo>
```
EOL

clbr-odoo commented on estate/views/estate_property_form_views.xml:
```diff
@@ -0,0 +1,26 @@
+        <field name="res_model">estate.property.offer</field>
+        <field name="view_mode">tree,form</field>
+    </record>
+    </odoo>
```
EOL

clbr-odoo commented on estate/views/estate_property_tag_views.xml:
```diff
@@ -0,0 +1,18 @@
+        <field name="view_mode">list,form</field>
+    </record>
+
+</odoo>
```
EOL

clbr-odoo commented on estate/views/estate_view_form.xml:
```diff
@@ -0,0 +1,73 @@
+            </form>
+        </field>
+    </record>
+</odoo>
```
EOL

clbr-odoo commented on estate/views/estate_view_form.xml:
```diff
@@ -0,0 +1,73 @@
+                    invisible="state in ['action_solde']"
+                         class="oe_highlight"/>
+                    <button name="action_cancel" string="CANCEL" type="object" 
+                    invisible="state in ['action_cancel']"
```
```suggestion
                    invisible="state == 'action_cancel"
```

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,97 @@
+    def _compute_best_price(self):
+        for rec in self:
+            prices = rec.offer_ids.mapped('price')
+            rec.best_price = max(prices) if prices else 0.0
```
That works, you can also do 
```suggestion
            rec.best_price = max(prices, default=0.0)
```

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,97 @@
+        "Best Offer", compute="_compute_best_price", readonly=True
+    )
+
+    @api.depends('living_area', 'garden_area')
+    def _compute_total_area(self):
+        for rec in self:
+            rec.total_area = (rec.living_area or 0) + (rec.garden_area or 0)
```
This compute is defined twice (see line 76).

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,97 @@
+            self.garden_orientation = False
+
+    def action_sold(self):
+        if "cancelled" in self.mapped("state"):
```
It works but it's not super readable. I'd do:
```suggestion
        if any([prop.state == "cancelled" for prop in self]):
```

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,97 @@
+    def action_sold(self):
+        if "cancelled" in self.mapped("state"):
+            raise UserError("Canceled property cannot be sold !")
+        return self.write({"state": "sold"})
```
```suggestion
        self.state = 'sold'
        return True
```
write doesn't return anything, and this notation also works, even if the recordset has many records.

clbr-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,48 @@
+    @api.depends('validity')
+    def _compute_date_deadline(self):
+        for offer in self:
+            if offer._origin.validity:
```
```suggestion
            if offer.validity:
```
validity will always be set, worst case to 0 so this if...else can be avoided.

clbr-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,48 @@
+            offer.state = 'accepted'
+            offer.property_id.selling_price = offer.price
+            offer.property_id.buyer_id = offer.partner_id.id
+            offer.property_id.state = 'offer_accepted'
```
```suggestion
            offer.property_id.write({
                'selling_price': offer.price,
                'buyer_id': offer.partner_id.id,
                'state': 'offer_accepted',
            })
```
It's better to do those three "write" in one :)

clbr-odoo reviewed (commented):
Don't forget to fix all missing EOL, keep up 👍

ybenbrai committed:
[IMP] estate: added acceptance criteria

	added acceptance criteria for the offers and the types, and
	colors..

ybenbrai committed:
[IMP] estate: Chapter 12: Inheritance

ybenbrai committed:
[FIX] estate: fixed warnings

	fixed label warning and constraint

ybenbrai committed:
[IMP] estate: added invoicing

	added invoicing debuged an issue when invoices does not appear

ybenbrai committed:
[IMP] estate: adding Kanban ViewKanban

	added kanban view and disabled DND, grouped the view

ybenbrai committed:
[FIX] estate: fixed spaces for linting

ybenbrai committed:
[FIX] estate: fixing lint and style issues

ybenbrai committed:
[FIX] estate: fixed the field hidden

ybenbrai committed:
[FIX] estate: fixing code guidance issue of field

ybenbrai committed:
[FIX] estate: fixing commits issues and added file
